text/0000-generic-closures.md

```diff
@@ -0,0 +1,104 @@
+- Feature Name: generic_closure
+- Start Date: 2015-06-15
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This RFC adds the ability to define closures that are generic over types.
+
+# Motivation
+[motivation]: #motivation
+
+Generic closures can be used to support compound operations on tuple types:
+
+```rust
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+struct Tuple<A, B, C>(pub A, pub B, pub C);
+
+impl<A, B, C> Tuple<A, B, C> {
+    fn map<A2, B2, C2, F>(self, mut f: F) -> Tuple<A2, B2, C2>
+        where F: FnMut(A) -> A2 + FnMut(B) -> B2 + FnMut(C) -> C2
+    {
+        Tuple(f(self.0), f(self.1), f(self.2))
+    }
+    
+    fn fold<T, F>(self, val: T, mut f: F) -> T
+        where F: FnMut(T, A) -> T + FnMut(T, B) -> T + FnMut(T, C) -> T
+    {
+        let val = f(val, self.0);
+        let val = f(val, self.1);
+        let val = f(val, self.2);
+        val
+    }
+}
+
+let a = Tuple(1u8, 2u32, 3.5f32).map(<T: Into<f64>>|x: T| x.into() + 1.0);
+assert_eq!(a, (2f64, 3f64, 4.5f64));
+
+let b = Tuple(1u8, 2u32, 3.5f32).fold(10.0, <T: Into<f64>>|x, y: T| x + y.into());
+assert_eq!(b, 16.5f64);
+```
+
+A fully working example of this code (with manually implemented closures) can be found [here](https://play.rust-lang.org/?gist=ea867336945253752e31873fc752ec06&version=nightly&backtrace=0).
+
+# Detailed design
+[design]: #detailed-design
+
+## Syntax
+
+There are two ways to specify generic bounds on closures:
+
+```rust
+<T: Debug>|x: T| println!("{:?}", x);
+
+<T>|x: T| where T: Debug {
+    println!("{:?}", x);
+}
+```
+
+When using the `where` syntax, the braces around the closure body are mandatory.
+
+If the `move` keyword is used then it must appear before the generic parameter list:
+
+```rust
+move <T: Debug>|x: T| println!("{:?}", x);
+
+move <T>|x: T| where T: Debug {
+    println!("{:?}", x);
+}
+```
+
+All generic parameters must be used in the closure argument list. This is necessary to ensure that the closure can implement all the required `Fn` traits.
+
+## Implementation
+
+The generated closure type will have generic implementations of `Fn`, `FnMut` and `FnOnce` with the provided type bounds. This is similar to the way closures currently have generic implementations over lifetimes.
```

Comment:
The way that closures are generic over lifetimes corresponds to types of the form `for <'a> Fn(&'a u8)` (for example), where the client code can instantiate that type at different lifetimes for `'a`.

Is that analogy what you intend to apply here, keeping in mind that the above lifetimes are erased at compile time?

I ask because when I first saw this proposal, I had assumed that part of the intention was to enable one to write [rank-N](https://wiki.haskell.org/Rank-N_types) higher-order functions (think `forall` in Haskell), where a function passed as an argument is able to be instantiated at multiple concrete types.

A concrete example of what I mean:

```rust
fn hof_example<F>(f: F) -> i32
    where F: for <T: Debug> Fn(T) -> i32
{
    f(0.0) + f(true)
}
```

Note: the above is certainly *doable* under a monomorphization strategy.
 * It becomes ... harder when one tries to generalize it to object-types (`f: Box<for <T: Debug> Fn(T)>`).

I'm just trying to understand the *scope* of what you're proposing.

----


In other words:

Can you speak more on the actual type assigned to these generic closures, or at least about the generic implementations here?

I.e. what type are you assigning to the expression 
```rust
<T: Debug> |x: T| { println!("{:?}, x); }
```
and what impls are generated for it?

I'm trying to figure out if this ends up being something like:

```rust
impl<T: Debug> Fn(T) for [closure@pnk-example] { ... }
```
or if there is something broader being implicitly proposed here.